packages/grid/x-data-grid-pro/src/hooks/features/detailPanel/useGridDetailPanelCache.ts

```diff
@@ -24,17 +26,16 @@ function cacheContentAndHeight(
     {},
   );
 
-  const heightCache = rowIds.reduce<Record<GridRowId, ReturnType<typeof getDetailPanelHeight>>>(
-    (acc, id) => {
-      if (contentCache[id] == null) {
-        return acc;
-      }
-      const params = apiRef.current.getRowParams(id);
-      acc[id] = getDetailPanelHeight(params);
+  const heightCache = rowIds.reduce<GridDetailPanelState['heightCache']>((acc, id) => {
+    if (contentCache[id] == null) {
       return acc;
-    },
-    {},
-  );
+    }
+    const params = apiRef.current.getRowParams(id);
+    const height = getDetailPanelHeight(params);
+    const autoHeight = height === 'auto';
+    acc[id] = { autoHeight, height: !autoHeight ? height : previousHeightCache[id]?.height };
```

Comment:
Nitpick: for better readability
```suggestion
    acc[id] = { autoHeight, height: autoHeight ? previousHeightCache[id]?.height : height  };
```